### ISSUE TITLE
ConcurrencyCheck support only valid when in update column

### ISSUE THREAD

**Issue by wangronghua** [dotnet/efcore#28909]
support a option to set if ConcurrencyCheck valid when ConcurrencyCheck column is in update column，that can reduce DbUpdateConcurrencyException when update other column

``` c#
public class EntityA
{
     public int Id{get;set;}
     public string Mobile{get;set;}
        [ConcurrencyCheck(OnlyWhenUpdate=true)]//OnlyWhenUpdate default false
     public int Score{get;set;}
}

Add(new EntityA{
Id=1,
Score=3
});

//not update socre,don't need score ConcurrencyCheck
var a=Load(1);
a.Mobile="xxxx";
SaveChanges();//  update EntityA set mobile='xxxx' where id=1

//update score need score ConcurrencyCheck
var a=Load(1);
a.Score=55;
SaveChanges();//  update EntityA set score=55 where id=1 and score=3
```

**Comment by roji** [comment]
@wangronghua the above is very much by design: the point of a concurrency token is usually to avoid modifying a database row if the value if that token changed since the row was loaded; this is unrelated to which property you're trying to change, especially since column values may be interrelated in the same row.

[The new ExecuteUpdate API](https://docs.microsoft.com/en-us/ef/core/what-is-new/ef-core-7.0/whatsnew#executeupdate-and-executedelete-bulk-updates) in EF Core 7.0 allows expressing any update query you want with LINQ, so you'd be able to use that API to do this.

**Comment by wangronghua** [comment]
Some concurrency controls are only used to control a single field from being overwritten by other writing processes or overwriting other writing processes. As I said above, the user can choose the option (onlywhenupdate = true) at this time. Imagine a scenario: the system is executing the fraction calculation service in the background. At this time, the user modifies the mobile. The user cannot modify the mobile because of the fraction calculation service in the background and the concurrent update fails. Executeupdate is a batch update that destroys the state of objects in memory. If only one piece of user data is updated, I think executeupdate should be used less

**Comment by roji** [comment]
> Executeupdate is a batch update that destroys the state of objects in memory. If only one piece of user data is updated, I think executeupdate should be used less

I'm not quite sure what this means... My point was that you can use ExecuteUpdate to achieve what you're asking for, i.e. update a column only if it has a certain value. For example, the following changes Foo to 8, but only if it's equal to 5.

```c#
ctx.Blogs.Where(b => b.Id == 1 && b.Foo == 5).ExecuteUpdate(s => s.SetProperty(b => b.Foo, b => 8));
```

**Comment by wangronghua** [comment]
> > Executeupdate is a batch update that destroys the state of objects in memory. If only one piece of user data is updated, I think executeupdate should be used less
> 
> I'm not quite sure what this means... My point was that you can use ExecuteUpdate to achieve what you're asking for, i.e. update a column only if it has a certain value. For example, the following changes Foo to 8, but only if it's equal to 5.
> 
> ```cs
> ctx.Blogs.Where(b => b.Id == 1 && b.Foo == 5).ExecuteUpdate(s => s.SetProperty(b => b.Foo, b => 8));
> ```

Thank you for your answer. I'm not sure about the features of executeupdate. If the local data has load before execution, will the local corresponding fields change after executeupdate is executed? In addition, the status of the local item is changed or modified

``` cs
var blog = ctx.Blogs.Find(1);

... do something

ctx.Blogs.Where(b => b.Id == 1 && b.Foo == 5).ExecuteUpdate(s => s.SetProperty(b => b.Foo, b => 8));

... use blog.Foo do something   //blog.Foo is 5 or 8?
ctx.Entry(blog) state is Unchanged or Modified
```

**Comment by roji** [comment]
No, ExecuteUpdate doesn't interact with change tracking, and any tracked entity (e.g. the blog loaded above with Find) won't be modified in any way. However, if your goal really is to update only one piece of user data (as indicated above), it sounds like you may not need change tracking... Note that ExecuteUpdate returns the number of rows affected, which tells you whether your update succeeded or not; that's very similar to the concurrency token approach in optimistic concurrency.

In other words, you can simply load your blog with a non-tracking query, and then use ExecuteUpdate as above to update the single column change.

**Comment by wangronghua** [comment]
I think this design is not very reasonable. In this way, what is tracking and what is non tracking in the system will be confused. When the non tracking object is passed to other module functions for use, the personnel of other modules also need to know the state of this object. Therefore, I think executeupdate should be used as little as possible, especially when only one record is changed, it will make the state of the object loaded locally inconsistent. In order to solve the problem in my above example, I can only split the score into a separate table to avoid the failure of mobile update due to the update of score, but this is not very elegant. If concurrencycheck can have an onlywhenupdate option, it will perfectly solve this problem.

**Comment by roji** [comment]
@wangronghua EF already supports both tracking and non-tracking queries and other operations; so yes, it's the programmer's responsibility to make sure they know whether they're using a tracking API or not. So this aspect of EF is by design, and I don't think ExecuteUpdate introduce anything problematic here.

Stepping back, the feature you are asking for could indeed limit concurrency exceptions to when only a specific property is being modified, and so reduce the number of exceptions. However, the idea of of optimistic concurrency in general is that concurrent updates of the same row are supposed to be rare, e.g. occuring only when two users happen to be editing the same row. Since your code has to deal with concurrency exceptions in any case, this would only reduce the frequency of those exceptions, which are again supposed to be rare in the first place. Where do you see the value of this, to the point where you're willing to modify the schema in order to do it?

**Comment by wangronghua** [comment]
@roji thank you for your answer. Yes, my change only reduces the probability of concurrent exceptions. It is true that this is a low probability event. In the actual scenario, there are few concurrent conflicts. However, it seems that optimistic lock conflicts can only prompt the user that the operation has failed. Pessimistic locks do not report an error without deadlock. If my modification will not affect the performance and stability of EF, I think the design will be more perfect. After all, it is a bad user experience to prompt the user that the operation fails. The probability can be reduced to as small as possible

**Comment by ajcvickers** [comment]
@wangronghua This is the first time I have seen a request to use concurrency checks in this way; it's certainly not a common design. I doubt this is something we will implement.

> It is a bad user experience to prompt the user that the operation fails.

There is no need to prompt the user if you determine that the concurrency failure is something that can be handled automatically.

**Comment by wangronghua** [comment]
@ajcvickers If the background service modifies the score and causes user modify the phone number concurrently fail, this can be solved automatically. However, if the background service modifies an individual field and causes user modify the score fail, this cannot be solved automatically. Even if it can be implemented without prompting the user for an error, it needs to write special code to implement it. This code will be written everywhere

**Comment by wangronghua** [comment]
Concurrencycheck is to prevent overwriting other process's data or being overwritten by others. This option only reduces the granularity of concurrency. Concurrencycheck is set on the score field. Then, the failure of the user to modify the mobile is caused by the fact that our background service is calculating his score. This is a little unreasonable. It is still reasonable for the user to modify the score. Besides, this option is not enabled by default. To enable this option, you must know the meaning of this option in advance

**Comment by roji** [comment]
> Even if it can be implemented without prompting the user for an error, it needs to write special code to implement it. This code will be written everywhere

Right, but that special code still needs to be there - everywhere - since your proposal only reduces the number of concurrency exceptions, and doesn't eliminate them. So this really has almost no value when the concurrency failure can be handled automatically.

Reading your last message, I'm still not quite sure how this feature would help you... It sounds like you have an automated, background service which occasionally updates a score column, but you want to allow users to manually update it (overwriting whatever the service calulates). If that's the case, then you can simply have a regular rowversion concurrency token, that's automatically updated when the row changes. If the user does a change and you get a concurrency exception (i.e. the automated service updated the score field right when the user was editing the row), you can simply reload the row and re-apply the user's changes, thereby overwriting the automated service's change. Since this is an automated handling strategy, there's no real advantage in making the concurrency check more granual for the score, as I wrote above. Let me know if I've someone misunderstood your scenario.

**Comment by wangronghua** [comment]
The example I gave earlier may not be very accurate. If the customer table has two columns: amount and score, the background periodically calculates the user's score, and the user's amount will be deducted when the user purchases goods. I think that the deduction of the user's amount for purchasing goods should not fail because the background service updates the score, and if you want not to throw this concurrent error, the existing code can't simply automatically process this error at this time, I need to write extra code to handle this error. There are two ways to avoid the failure of the amount update caused by the score update: one is to separate the score and the amount into two tables, so that optimistic locks will not conflict, but it is not elegant enough, and the table structure needs to be changed. The second method is the concurrencycheck (onlywhenupdate = true) I mentioned here, which can avoid concurrency errors caused by this situation, and it is very simple. I don't know if it will affect the underlying performance and stability of EF. If it doesn't, I think it is a good method

**Comment by roji** [comment]
Have you considered simply defining amount as your concurrency token, instead of having an auto-generated rowversion? That would mean that updates will only fail if the amount isn't the same as it was when the row was loaded. This is quite similar to what you're trying to achieve above: it means that updates to amount wouldn't fail just because score changed (since the update SQL would only contain `WHERE amount = <original_amount>`. It does mean that your score-calculating service would get concurrency exceptions, but these are supposed to be easy to manage automatically etc.

**Comment by wangronghua** [comment]
Defining rowversion has the largest granularity and is most prone to concurrency conflicts. Concurrent updates of any two fields will lead to failure. In general, I will avoid using rowversion. Defining amount and score as concurrency token granularity is smaller than rowversion, but they will also lead to conflicts. To solve this conflict, I need to add onlywhenupdate. If only defining amount as concurrency token, In this case, the program has concurrent coverage of the score field. For example, the user uses the score and the background service calculates the user's score at the same time. There may be cases where the user uses the score but does not deduct it

**Comment by ErikEJ** [comment]
Wondering if the relational model could be improved?

**Comment by wangronghua** [comment]
@ErikEJ Both amount and score are attributes of the customer. I don't know how to improve them. They may have to be artificially split into two tables. They have a 1-to-1 relationship. This split table is simply to avoid the possible concurrency conflict between two concurrency fields in one table. If there are similar attributes later, another table may have to be created to avoid the concurrency conflict between them

**Comment by ajcvickers** [comment]
@wangronghua Assuming we added this feature, what code would you write to handle the thrown concurrency exception?

**Comment by wangronghua** [comment]
@ajcvickers If this feature is available, the amount update and the score update will not lead to concurrency errors. Only the amount update and the amount update will lead to concurrency errors, and only the score update and the score update will lead to concurrency errors. This concurrency error is unavoidable and required by the concurrency mechanism

**Comment by wangronghua** [comment]
@ajcvickers Oh, sorry to misunderstand what you mean. The simplest way to deal with it without reporting an error is to re create new dbcontext, and then re execute all the original logic according to the original logic

**Comment by ajcvickers** [comment]
@wangronghua Right, but to be a bit more concrete, can you show the code?

**Comment by wangronghua** [comment]
similar
``` c#
        private async Task ExecuteCore(Func<CurrencyServiceDbContext, Task> func)
        {
            var concurrencyErrorCount = 0;
            while (concurrencyErrorCount < 2)
            {
                try
                {
                    using (var context = new CurrencyServiceDbContext())
                    {
                        await func(context);
                    }
                    return;
                }
                catch (DbUpdateConcurrencyException e)
                {
                    concurrencyErrorCount++;
                }
            }
            throw new Exception("并发冲突，请重新操作");
        }





            //use like 
            await ExecuteCore(async context => {
                var entity = await context.Set<Customer>().FindAsync(1);
                if (entity.Score + score >= 0)
                {
                    entity.Score += score;
                    ...
                    await context.SaveChangesAsync();
                }
            });
```

**Comment by ajcvickers** [comment]
@wangronghua Thanks for sharing. I think this reaffirms @roji's point above that the current concurrency mechanism should work just fine with the code you posted. Occasionally an update to the score maybe sent from the service when the application didn't actually change the score, but this should not be detrimental in any way.

So, after discussing this with the team, we don't plan to make any changes in this area.

**Comment by wangronghua** [comment]
This is just a demo processing method. There will be several problems in the actual project if it is implemented in this way

1. Dbcontext is injected by scope, so it is difficult to create new

2. Even if the new dbcontext can be operated, the acquisition and modification of the model need to be in one function. The business logic cannot be reused, and duplicate code will be written everywhere,I think this is a very important feature

**Comment by ajcvickers** [comment]
Neither of those things change if we implement this feature.
